source/Cosmos.System2/Network/IPv4/TCP/FTP/FtpServer.cs

```diff
@@ -0,0 +1,180 @@
+        public string Content { get; set; }
+    }
+
+    /// <summary>
+    /// FtpServer class. Used to handle FTP client connections.
+    /// </summary>
+    public class FtpServer : IDisposable
+    {
+        /// <summary>
+        /// Command Manager.
+        /// </summary>
+        internal FtpCommandManager CommandManager { get; set; }
+
+        /// <summary>
+        /// TCP Listener used to handle new FTP client connection.
+        /// </summary>
+        internal TcpListener tcpListener;
+
+        /// <summary>
+        /// Is FTP server listening for new FTP clients.
+        /// </summary>
+        internal bool Listening;
+
+        /// <summary>
+        /// Create new instance of the <see cref="FtpServer"/> class.
+        /// </summary>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown on fatal error (contact support).</exception>
+        /// <exception cref="ArgumentException">Thrown if UdpClient with localPort 53 exists.</exception>
+        public FtpServer(CosmosVFS fs, string directory)
+        {
+            if (Directory.Exists(directory) == false)
+            {
+                throw new Exception("FTP server can't open specified directory.");
+            }
+
+            CommandManager = new FtpCommandManager(fs, directory);
+
+            Listening = true;
+        }
+
+        /// <summary>
+        /// Listen for new FTP clients.
+        /// </summary>
+        public void Listen()
+        {
+            while (Listening)
+            {
+                tcpListener = new TcpListener(21);
+                tcpListener.Start();
+                var client = tcpListener.AcceptTcpClient();
+
+                global::System.Console.WriteLine("Client : New connection from " + client.StateMachine.LocalAddress.ToString());
+
+                ReceiveNewClient(client);
+            }
+        }
+
+        /// <summary>
+        /// Handle new FTP client.
+        /// </summary>
+        /// <param name="ftpClient">FTP Client.</param>
+        private void ReceiveNewClient(TcpClient client)
+        {
+            var ftpClient = new FtpClient(client);
+
+            ftpClient.SendReply(220, "Service ready for new user.");
+
+            while (ftpClient.Control.IsConnected())
+            {
+                ReceiveRequest(ftpClient);
+            }
+
+            ftpClient.Control.Close();
+
+            //TODO: Support multiple FTP client connection
+            Close();
+        }
+
+        /// <summary>
+        /// Parse and execute FTP command.
+        /// </summary>
+        /// <param name="ftpClient">FTP Client.</param>
+        private void ReceiveRequest(FtpClient ftpClient)
+        {
+            var ep = new EndPoint(Address.Zero, 0);
+
+            try
+            {
+                var data = Encoding.ASCII.GetString(ftpClient.Control.Receive(ref ep));
+                data = data.Remove(data.Length - 2, 2);
+
+                global::System.Console.WriteLine("Client : " + data);
```

Comment:
Same here. In future it would be best if we could use events to notify the OS that something happened but I dont know if they are currently supported